csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs

```diff
@@ -619,51 +641,138 @@ public string EndProfiling()
+                var utf8Name = meta.ZeroTerminatedUtf8Name;
+                Debug.Assert(utf8Name != null);
+                var pinnedHandle = new Memory<byte>(utf8Name).Pin();
+                unsafe
+                {
+                    result[i] = (IntPtr)pinnedHandle.Pointer;
+                }
                 cleanupList.Add(pinnedHandle);
             }
             return result;
         }
 
         /// <summary>
         /// This function obtains ortValues for NamedOnnxValue.
-        /// The problem with NamedOnnxValue is that it does not contain any Onnx (OrtValue)
-        /// so calling ToOrtValue creates a new instance of OrtValue that needs to be disposed.
+        /// The problem with NamedOnnxValue is that it is not disposable and can not contain any disposable items.
+        /// so calling InputToOrtValue creates a new instance of OrtValue that needs to be disposed.
         /// The deriving object DisposableNamedValue actually contains and owns OrtValue and it returns
         /// it.
         /// </summary>
-        /// <param name="values"></param>
-        /// <param name="cleanupList"></param>
+        /// <param name="values">a collection of NamedOnnxValues</param>
+        /// <param name="metaLookup">Metadata lookup function (input/initializers/output)</param>
+        /// <param name="cleanupList">list to cleanup in an exception safe manner</param>
         /// <returns></returns>
-        private IntPtr[] GetOrtValuesHandles(IReadOnlyCollection<NamedOnnxValue> values, DisposableList<IDisposable> cleanupList)
+        private IntPtr[] GetOrtValuesHandles(IReadOnlyCollection<NamedOnnxValue> values, MetadataLookup metaLookup,
+            OrtValueExtractor ortValueExtractor,
+            DisposableList<IDisposable> cleanupList)
         {
+            cleanupList.Capacity += values.Count * 2;
             IntPtr[] result = new IntPtr[values.Count];
-            for (int inputIndex = 0; inputIndex < values.Count; ++inputIndex)
+            for (int valueIndex = 0; valueIndex < values.Count; ++valueIndex)
             {
-                var input = values.ElementAt(inputIndex);
-                MemoryHandle? memHandle;
-                var ortValue = input.ToOrtValue(out memHandle);
-                if (memHandle.HasValue)
+                var value = values.ElementAt(valueIndex);
+                var meta = metaLookup(value.Name);
+                var ortValue = ortValueExtractor(value, meta, out IDisposable memHolder);
+                if (memHolder != null)
+                {
+                    cleanupList.Add(memHolder);
+                }
+                if (ortValue != null)
+                {
+                    if (ortValue.IsOwned)
+                        cleanupList.Add(ortValue);
+
+                    result[valueIndex] = ortValue.Handle;
+                }
+                else
                 {
-                    cleanupList.Add(memHandle);
+                    result[valueIndex] = IntPtr.Zero;
```

Comment:
>result[valueIndex] = IntPtr.Zero; [](http://example.com/codeflow?start=20&length=33)

NamedOnnxValue instances can not create output structures for sequences and maps beforehand, as it is done for a tensor buffer. So it returns zero ptr.
Ort then creates output OrtValues.
TODO: We need to create output structures for those and dispose.

Reply:
NamedOnnxValue now throws when an ort value is created on top of non-tensors for output.